File: getting-a-different-number.py
# Getting a Different Number

# Given an array arr of n unique non-negative integers, how can you most 
# efficiently find a non-negative integer that is not in the array?

# Your solution should return such an integer or null if arr contains all 
# possible integers.

# Solution 1: Use pigeonhole principle and hashing. Create an array arr2 of 
# length n+1. Use hash function to map each number in arr to an index of arr2.
# From pigeonhole principle at least on cell will be empty. Hash function can be
# f(x) = x mod (n+1). For any index j of an empty cell in arr2, j is not in arr.

# Any integer j + m(n+1) for any m in [0, MAX_INT] is a number not in arr.

function getAnotherNumber(arr):
	n = len(arr)
	if n == MAX_INT: return null
	arr2 = [-1] * (n + 1)
	for i in arr:
		arr2[i % (n+1)] = i
	for i in arr2:
		if i == -1: return i

# Solution 2: Coming up with any n+1 numbers. 
# Since arr has n numbers, scanning n+1 numbers would give one number that is not
# in arr. To determine if a number is in arr or not in O(1), we can use hash.
# Hash key is number from arr, hash value is unimportant

function getAnotherNumber1(arr):
	n = len(arr)
	if n == MAX_INT: return None
	dict = {}
	for i in a:
		dict[i] = 1
	for i in range(1, n+2):
		if dict[i] == None:
			return i
	return None # If we reach this line arr has more than n unique numbers
	
	

File: getting-a-different-number.java
/* package whatever; // don't place package name! */

import java.util.*;
import java.lang.*;
import java.io.*;

/* Name of the class has to be "Main" only if the class is public. */
class Ideone
{
	public static void main (String[] args) throws java.lang.Exception
	{
		int[] arr = new int[];
		System.out.println(getAnotherNumber1(arr));
		System.out.println(getAnotherNumber2(arr));
	}
	
	public static int getAnotherNumber1(int[] arr) {
		int n = arr.length;
		if (n == Integer.MAX_VALUE+1) return null;
		HashSet<Integer> set = new HashSet<Integer>();
		for (int i = 0; i < n; i++) {
			set.offer(arr[i]);
		}
		for (int i = 0; i <= n; i++) {
			if (!set.contains(i))
				return i;
		}
		return null;
	}
	
	public static int getAnotherNumber2(int[] arr) {
		int n = arr.length;
		if (n == Integer.MAX_VALUE+1) return null;
		int[] arr2 = new int[n+1];
		for (int i = 0; i < n; i++) {
			arr2[arr[i] % (n+1)] = arr[i];
		}
		for (int i = 0; i < n+1; i++) {
			if (arr2[i] == 0) 
				return i;
		}
	}
}
